### PR TITLE
Added extra experiments - mainly around macro chunking

### PR DESCRIPTION
**Pull request by dannyjameswilliams**
## Overview 
Experiments added:
 - LongEmbed Examples against chunk size (nDCG@10 and mAP@10)
 - Macro chunking approach vs 'hard' boundary approach with 0 overlap
 - Example with Anthropics contextual retrieval

### LongEmbed Examples against chunk size (nDCG@10 and mAP@10)
Similarly to `run_chunked_eval.py`, `run_chunked_eval_with_macro_chunking.py` can just be run on the command line with e.g.
```bash
python3 run_chunked_eval_with_macro_chunks.py --task-name LEMBWikimQARetrievalChunked
```

## To reproduce easily

I recommend the bash file
```bash
#!/bin/bash

# Define arrays for names and embedding sizes
names=(LEMBWikimQARetrievalChunked LEMBQMSumRetrievalChunked LEMBNarrativeQARetrievalChunked LEMBSummScreenFDRetrievalChunked)

# Loop over each name
for name in "${names[@]}"; do
  echo $name
  python3 run_chunked_eval_with_macro_chunks.py --task-name $name
  done
```

to run them all at once. Then the results can be displayed graphically in a matplotlib plot via running the file `plot_chunk_size_experiments.py`.

### Macro chunking approach vs 'hard' boundary approach with 0 overlap
Similar to the above - comparing macro chunking to non-macro chunking, with experiment file `run_macro_chunking_experiments.py` and plot file `plot_macro_chunking_experiments.py`.

### Example with Anthropics contextual retrieval

You can run the `explanatory_contextual_retrieval.py` to see a comparison between Anthropics contextual retrieval, which manually adds context to each chunk, late chunking, and naive chunking. This is performed via a running on a generated document which deliberately has context missing in later sentences (via 'Its' instead of a company name).  The comparison is via cosine similarities on the chunks and the corresponding embeddings on `jina-embeddings-v2-base-en`.

### CODE REVIEW

File: chunked_pooling/chunked_eval_tasks.py [hunk 1]
```diff
@@ -297,6 +297,245 @@ def load_data(self, **kwargs):
         self.data_loaded = True
 
 
+class LEMBNarrativeQARetrievalChunked(AbsTaskChunkedRetrieval):
```
Comment (guenthermi) [inline chunked_pooling/chunked_eval_tasks.py]: As we have already another version of it, I would suggest to remove this one

File: chunked_pooling/chunking.py [hunk 2]
```diff
@@ -150,8 +150,8 @@ def chunk(
                 tokenizer=tokenizer,
             )
         elif chunking_strategy == "fixed":
-            if chunk_size < 10:
-                raise ValueError("Chunk size must be greater than 10.")
+            if chunk_size < 8:
+                raise ValueError("Chunk size must be >= 8.")
```
Comment (guenthermi) [inline chunked_pooling/chunking.py]:
```suggestion
            if chunk_size < 4:
                raise ValueError("Chunk size must be >= 4.")
```

File: plot_chunk_size_experiments.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import json
+import os
+
+if __name__ == "__main__":
```
Comment (guenthermi) [inline plot_chunk_size_experiments.py]: I would remove the scripts for the plots as the plots in the paper are generated with different scripts.

File: plot_macro_chunking_experiments.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
+import numpy as np
+import matplotlib.pyplot as plt
+import json
+import os
+
+if __name__ == "__main__":
```
Comment (guenthermi) [inline plot_macro_chunking_experiments.py]: same here

File: run_chunked_eval_with_macro_chunks.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        #     **chunking_args,
+        # )
+        # evaluation.run(
+        #     model,
+        #     output_folder=f'results-chunked-pooling/chunk_size_{chunk_size}',
+        #     eval_splits=[eval_split],
+        #     overwrite_results=True,
+        #     batch_size=BATCH_SIZE,
+        #     encode_kwargs={'batch_size': BATCH_SIZE},
+        # )
```
Comment (guenthermi) [inline run_chunked_eval_with_macro_chunks.py]: I think this script needs a bit of a clean up, maybe we can also integrate it into the main script

File: run_macro_chunking_experiments.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    strategy,
+    task_name,
+    eval_split,
+    chunking_model,
+    truncate_max_length,
+    chunk_size,
+    n_sentences,
+    soft_boundary_embed_size,
+    hard_boundary_embed_size,
+):
```
Comment (guenthermi) [inline run_macro_chunking_experiments.py]: As we didn't run this experiment, I think it does not make sense to merge this script

**Review by guenthermi — COMMENTED** [review]
Thank you for contributing all the code, I added some comments. Would be nice if you could address them before we merge this

File: run_chunked_eval_with_macro_chunks.py (hunk 5 above)
Comment (dannyjameswilliams) [reply run_chunked_eval_with_macro_chunks.py]: I have integrated into the main script and also renamed macro chunking to long late chunking. I've got long late chunking off by default but when defining e.g. `--long-late-chunking-embed-size 8192`, hope this is what you meant

**Comment by dannyjameswilliams** [comment]
all comments hopefully addressed

File: explanatory_contextual_retrieval.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    def __init__(self, 
+            model: AutoModel,
+            tokenizer: AutoTokenizer, 
+            llm_name: str = "meta-llama/Meta-Llama-3.1-8B",
+            chunking_strategy: str = "fixed"
+        ):
+        # self.llm = pipeline(
+        #     "text-generation", model=llm_name, model_kwargs={"torch_dtype": torch.bfloat16}, device_map="auto",
+        #     max_length = 1000
+        # )
```
Comment (guenthermi) [inline explanatory_contextual_retrieval.py]:
```suggestion
```

File: explanatory_contextual_retrieval.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
+# experiments/explanatory_contextual_retrieval.py
+# 
+# a simple example with a trivial piece of text to showcase the late chunking method against 
+# contextual retrieval method. contextual retrieval manually inserts context to each
+# chunk, i.e. forces context to be around each chunk. so works as a good comparison
+# to late chunking to see if the similarities are similar (which they appear to be)
+# 
+# pip requirements:
+# accelerate?
```
Comment (guenthermi) [inline explanatory_contextual_retrieval.py]: can you  add this to the pyproject.toml if necessary?

File: explanatory_contextual_retrieval.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+    text = """
+    The recent SEC filing provided insights into ACME Corp's performance for Q2 2023. 
+    It highlighted a 3% revenue growth over the previous quarter. 
+    The company, which had a revenue of $314 million in the prior quarter, showed steady progress. 
+    They attributed this growth to strategic initiatives and operational efficiencies. 
+    The report emphasized the company's resilience and ability to navigate market challenges, reflecting positively on their financial health and future prospects.
+    """.strip().replace("\n", "")
+
+
+    # llm_model_name = "microsoft/Phi-3.5-mini-instruct"
```
Comment (guenthermi) [inline explanatory_contextual_retrieval.py]:
```suggestion
```

**Review by guenthermi — APPROVED** [review]
just added some minor comments

**Pull request merged**